microsoft/onefuzz
# Refactor AdoTemplate rendering

## Summary of the Pull Request

_What is this about?_

Closes #3369 
Closes #1692

## Info on Pull Request

_What does this include?_

* A new model `RenderedAdoTemplate`
* Once per notification attempt, we render our `AdoTemplate` into a `RenderedAdoTemplate`. During that process, we can enforce limitations imposed by ADO such as a max title length
* Later, when using the fields from the config, we don't need to remember/worry about re-rendering or making sure the rendered results meets the ADO limitations
* Refactor `AdoConnector` to only work with a `RenderedAdoTemplate` to enforce the behavior from the previous 2 bullet points 
* `AdoConnector` no longer has the ability to render fields in a config
* Split `ExistingWorkItems` into 2 parts: - Allowing us to test the query we generate
    1. Creating the ADO query
    2. Executing the query and returning results
* A test to validate the ADO query we're generating
* A test to validate that the title is truncated when rendering `AdoTemplate` to `RenderedAdoTemplate`

## Validation Steps Performed

_How does someone test & validate?_

Validated in test instance:

1. Start a new job
2. Attach a notification config to the job and make the config so that it's guaranteed to truncate the title (I just repeated report.crash_site over and over)
3. Create a debug notification for the job from step 1
4. Verify a bug was created with the appropriate template and that the title was truncated
5. Repeat step 3 with a new crash name
6. Verify the bug from step 4 was updated with a comment containing the new crash name
7. Verify a new bug was not created
